### DISCUSSION TITLE
$inertia.post() results in a 302 redirect.

### DISCUSSION THREAD

**Discussion by vannut** [inertiajs/inertia#967] [Help] [answered]
I've got a small inertia app which gets loaded into another domain through an Iframe. 

If i want to update something server side I make a request in the vue-component via:
```vue
this.$inertia.post(
  `/add-to-cart?ulid=${this.ulid}`,
  { test: 1234 }   
);
```

Serverside, in my controller I perform some logic en redirect the request to a second page:

```php
return \Redirect::route('cart', [
            'ulid' => $request->get('ulid')
])
->withMessage([
    'label' => 'bijgewerkt',
    'type' => 'cart-bijgewerkt'
]);
```

If I visit the app on its own domain (so not iframed) this redirect works like expected. I end up on the `cart` route and there is a flash message.

But when the app is being iFramed, it does not work. 
The logic (adding an item to the cart) is not being done, I dont see an item in the Database.
And If I look into the network tab I see the an XHR request being performed, with the correct POST data. But it has a response code of **302**

Even when I just replace the whole controller method with a `dd('hello')` I would expect it to show up... but a 302 is again performed.

I'm a little bit lost on why this redirect is happening or even where it comes from!

**Comment by ajnsn** [comment, marked as answer]
This is probably a Token Mismatch Redirect because of the "Lax" SameSite policy. The XSRF-TOKEN cookie cannot be accessed when posting to your server via an Iframe.

Jonathan has sum up some more information here: https://inertiajs.com/csrf-protection. You can try to share the token as a prop and then include it in the post data, as mentioned in the beginning of the article.

There may be also other solutions availabe. But they require a more deep understanding what SameSite cookies are and how you need to configure them to keep your app secure.

**Reply by vannut** [reply]
Thanks! That pointed me in the correct directen.
I would have thought to get the dreaded "419 Token Mismatch" exception then. But apperently it just sends a 302.. 

Digging a little deeper with your guidance I found that the  xsrf-cookie wasnt even set because of Lax/Strict setting of the session.

Changing the `session.same_site` to `none` resulted in the the session and xscrf-cookie to get set. And subsequentially the  `this.$inertia.post()`  picks up the xsrf cookie and we got a working redirect again. 

I'm uncertain what security implications the change of the 'session.same_site'setting has.

**Reply by ajnsn** [reply]
Setting SameSite to "none" will only work in a secure (SSL) environment. Here's a good [explanation about the topic](https://web.dev/samesite-cookies-explained/). The security implication is most likely a Cross-site request forgery, so every host can trigger requests to your app and your browser will attach the cookie. You could also take a look at the [X-Frame-Options ](https://developer.mozilla.org/docs/Web/HTTP/Headers/X-Frame-Options) to narrow down the hosts allowed iframing your app.

**Reply by vannut** [reply]
Thanks for your time @ajnsn, really appreciate it!

After reading into the SameSite: none/lax/strict settings. It appears all problems should be solved if set to `none`. It even recommends it to set it to none:
![Schermafbeelding 2021-11-12 om 08 40 31](https://user-images.githubusercontent.com/7419859/141428927-98b086de-b309-4167-8aaf-ae56dbe6617e.png)

But there are some implementation problems with older browsers on how they handle none; making it not a viable solution. 

If I exclude the route from csrf-checks in the `VerifyCsrfToken` the behaviour is as expected: it works. But obviously without the csrf-protection.

Looking at the inertia page on handeling csrf protection I followed te csrf-in-shared-props-route.
Adding it to my InertiaMiddleware:
```php
public function share(Request $request)
{
  return array_merge(parent::share($request), [
    'csrf_token' => csrf_token(),
  ]);
}
```
And then changing the post-function to include the csrf token:

```javascript
handleClick() {
  console.log(`from props: ${this.$inertia.page.props.csrf_token}`);
  this.$inertia.post(
    `/add-to-cart?ulid=${this.ulid}`,
    { 
      test: 1234,
      _token: this.$inertia.page.props.csrf_token
    }
  )
},
```

This is not working.
Even worse: looking at the console it appears the token is being reset on every click:
![Schermafbeelding 2021-11-12 om 08 48 27](https://user-images.githubusercontent.com/7419859/141429917-8176c77c-d4a2-489e-9aa6-c8ad29074ea8.png)

So after thinking of understanding the problem, I'm completly lost again :(

**Reply by ajnsn** [reply]
The token will be generated for each request, that's the desired behaviour.
<strike>I think you should use this.$page.props.csrf_token instead of this.$inertia.page.props.csrf_token?</strike>

Edit: Just recognized that both ways to access the props work :) 

I tried your approach and for me this works, e.g. 
```js
submit() {
    this.$inertia.post('profile', {
        name: 'Jonathan',
        email: 'jonathan@reinink.ca',
        _token: this.$page.props.csrf_token,
    });
},
```

The only possible thing different is that I added `protected $addHttpCookie = false;` to the `VerifyCsrfToken` middleware as I need to simulate that there's no `XSRF-TOKEN`.

**Reply by vannut** [reply]
> The token will be generated for each request, that's the desired behaviour.

If it's the same session it should be the same right? Otherwise the server does not know the correct token to check it against.  If I visit the iframed page outside the iframe the token remains the same over subsequent refreshes.

Just to confirm: you tried it inside an iframe, with samesite on lax?
I added the addHttpCookie to my middleware: no joy :(

**Reply by ajnsn** [reply]
Haven't tested in an iframe but your issue might now be the missing session-cookie, not the CSRF protection itself. To prove this, you could set `session.secure` to `true` and `session.same_site` to `none`. Then add `protected $addHttpCookie = false;` to the `VerifyCsrfToken` middleware. You should then only receive the session-cookie inside the iframe but the CSRF protection will be handled by sending over using the shared prop. Might be not a direct solution but session handling is another topic.

**Reply by vannut** [reply]
All right it seems that the session also must be stored with a cookie. Which because of the samesite policy gets rejected. And as the csrf token is bound to the session, it gets resetted on every request. 

Though setting `samesite=none` would mitigate this and let the cookie to be set; safari on macOS 10.15 / catalina won't properply process this setting.

So in the end, I just don't use session management.

**Reply by ajnsn** [reply]
Oh dear, yes. Other options would be to use stateless tokens, like JWT or even transfering/exchanging the session ID within the headers and put it into localStorage. But this removes a lot of the simplicity that Inertia provides out of the box, so in the end iframes keep beeing complicated, maybe you can avoid them somehow.

**Reply by vannut** [reply]
The use case is a simple embed of (part of) my application a clients website. So I'm actually bound to use an iframe :) Thanks for your feedback!
